### ISSUE TITLE
Rust and Python implementation treat cleanup differently

### ISSUE THREAD

**Issue by sebpuetz** [huggingface/tokenizers#340]
Hi,

```Py
>>> tokenizer = transformers.AutoTokenizer.from_pretrained("bert-base-multilingual-cased", use_fast=True)
>>> [tokenizer.tokenize(t) for t in ['Test', '\xad', 'test']]
[['Test'], [], ['test']]

>>> tokenizer = transformers.AutoTokenizer.from_pretrained("bert-base-multilingual-cased", use_fast=False)
>>> [tokenizer.tokenize(t) for t in ['Test', '\xad', 'test']]
[['Test'], ['[UNK]'], ['test']]
```

Digging a bit in the code, `tokenizers::normalizers::bert::is_control` returns `true` for `SOFT-HYPHEN (\xad)`, so it gets filtered in `tokenizers::normalizers::bert::BertNormalizer::do_clean_text`.

It would be nice if the Rust implementation could also return `'[UNK]'` when it removes all characters of a token. While having a control character in the dataset is clearly an issue with the data, I still think consistency would be nice.

**Comment by sebpuetz** [comment]
If I'm not mistaken, the Python implementation of the BERT tokenizer doesn't clean / normalize its inputs. Therefore it passes `'\xad'` to its `tokenize()` method where it gets `UNK`ed as a proper unknown token. I can replicate this behaviour by setting `clean_text` to `False` in `BertTokenizerFast`'s init.


```Py
>>> tokenizer = transformers.AutoTokenizer.from_pretrained("bert-base-multilingual-cased", use_fast=True, clean_text=false)
>>> [tokenizer.tokenize(t) for t in ['Test', '\xad', 'test']]
[['Test'], ['[UNK]'], ['test']]
```

**Comment by n1t0** [comment]
Thank you for reporting this! There is indeed a difference that must be addressed, but it is originating in the Python implementation that shouldn't return the `[UNK]` token, in order to say consistent with the original implementation. This is being fixed in `transformers` [here (#5749)](https://github.com/huggingface/transformers/pull/5749)

**Comment by sebpuetz** [comment]
If that's the default behaviour, then there's a different issue with the Rust implementation which I'm not sure how to address:

I'm working with pre-tokenized data (UD-Treebanks) for a sequence-tagging task, since I don't want to inflate the importance of a training example based on the number of word-pieces the token gets split into, I need to map the labels to only the first word-piece of a token. 

To achieve this, I was iterating over the words in the original sentence as taken from the treebank and used the `word_to_tokens` method with the offset of the word in the sentence to get the corresponding token span. If words simply vanish from the sentence, then at first the offsets become invalid and at the final word of the sequence an exception is raised because there's no offset for disappearing words in the sequence.

**Comment by sebpuetz** [comment]
The code snippet should make the issue clearer:

~~~Py
        encoded_batch = tokenizer.batch_encode_plus(
            [[token.form for token in sentence] for sentence in batch],
            padding=True,
            is_pretokenized=True,
            return_tensors='pt',
            truncation=True,
            max_length=truncate_len)
        first_pieces = torch.zeros_like(encoded_batch.attention_mask, dtype=torch.bool)

        for row, sentence in enumerate(batch):
            for col, token in enumerate(sentence):
                idx = encoded_batch.word_to_tokens(row, col)[0] # this method raises the exception
                first_pieces[row, idx] = True
~~~ 

The error message is also not clear in that example, it fails on a call to construct a `TokenSpan` while trying to unpack a `None` value. So it's a hard-to-debug issue, too.

**Comment by n1t0** [comment]
Can you try to extract a complete example to allow us to easily reproduce this, and open a new issue with it?
